### PR TITLE
[Core][RL] Support sparse checkpoint updates through native weight loaders

### PR DESCRIPTION
**Pull request by ShuoleiWang**
### Motivation

Online RL trainers commonly produce updates in checkpoint/Hugging Face coordinates: a checkpoint name, the full checkpoint shape, and either a dense tensor or flat changed indices and values. These coordinates are not the same as vLLM's runtime parameter coordinates.

For example, a checkpoint tensor may be split across tensor-parallel ranks, packed with another checkpoint tensor, or mapped into expert-parallel runtime weights. The existing `sparse_nccl` backend expects an already-resolved runtime parameter name and rank-local indices, so it cannot serve as a general bridge for checkpoint-coordinate updates.

This PR adds a small model-loader API that keeps the model's native `model.load_weights()` path as that bridge. The native loader continues to own checkpoint-name mapping, TP/EP placement, and packed/model-specific layouts.

### Design

```mermaid
flowchart LR
    A["One patch-loading call<br/>checkpoint name, full shape/dtype,<br/>values and optional flat indices"]
    A --> B{"All patches in this call"}
    B -- "Dense" --> C["Build full checkpoint tensors"]
    B -- "Sparse" --> D["Build full checkpoint-shaped NaN tensors<br/>index_copy_ fills changed checkpoint positions"]

    C --> L["One or more model.load_weights(...) calls<br/>native checkpoint-to-runtime mapping<br/>for TP/EP and packed/model-specific layouts"]
    D --> P["Temporarily replace process-wide<br/>torch.Tensor.copy_<br/>during each serialized sparse loader call"]
    P --> L

    L -- "Dense: native loader semantics" --> U["Runtime tensors updated"]
    L -- "Sparse: each Python Tensor.copy_" --> E{"Floating source and destination<br/>with the same shape?"}
    E -- "Yes" --> M["Cast source, merge with<br/>where(isnan(source), destination, source),<br/>then call the original copy_"]
    E -- "No" --> N["Call the original copy_ unchanged<br/>unsupported for sparse patches"]
    L -. "Sparse custom or non-copy_ write" .-> X["Not intercepted<br/>unsupported for sparse patches"]

    M --> U
    N --> Y["Outside the supported sparse behavior"]
    X --> Y
    P -. "finally, after model.load_weights<br/>returns or raises" .-> R["Restore the original Tensor.copy_"]
```

`CheckpointWeightPatch` represents either:

- a dense checkpoint replacement (`indices=None`), or
- a sparse checkpoint patch (`indices` and `values` over the full checkpoint shape).

Sparse patches are expanded into checkpoint-shaped NaN staging tensors. The model's native loader maps those tensors into runtime-local destinations. For the supported single-copy loader path, the final same-shaped floating-point `Tensor.copy_` keeps the existing destination at NaN positions and copies the changed values elsewhere.

Patches are validated before the first load. Loader calls are bounded by a checkpoint-byte batching target, and a repeated checkpoint name begins a new loader call so repeated patches preserve input order.

### Scope and failure semantics

The current implementation is intentionally narrow:

- sparse patches require a floating checkpoint dtype and cannot carry NaN as a real update value;
- the loader must preserve NaNs through checkpoint-to-runtime mapping and end in a same-shaped floating-point `Tensor.copy_`;
- the validated integration uses exclusive, serialized weight updates and a single-copy loader;
- composed/multi-stage loaders and custom write paths are not supported;
- sparse application is not O(nnz): it allocates a full checkpoint-shaped NaN tensor and runs a dense `torch.where`/copy at the runtime destination;
- updates are not transactional. A loader error can leave a partial update, so the caller must restore a known baseline or restart and reseed before serving.

### Companion VERL integration

[VERL #7227](https://github.com/verl-project/verl/pull/7227) uses this API to add a vLLM consumer for VERL's `delta_sharded` checkpoint-engine wire format. The first update is a dense seed; steady updates contain checkpoint-coordinate indices and changed values.

### Duplication check

I searched open vLLM PRs for checkpoint-coordinate weight patch loading and found no matching implementation.

## Test Plan

```bash
.venv/bin/python -m pytest \
  --confcutdir=tests/model_executor/model_loader \
  tests/model_executor/model_loader/test_checkpoint_weight_patch.py -q

.venv/bin/ruff check --no-cache \
  vllm/model_executor/model_loader/checkpoint_weight_patch.py \
  tests/model_executor/model_loader/test_checkpoint_weight_patch.py

.venv/bin/ruff format --check --no-cache \
  vllm/model_executor/model_loader/checkpoint_weight_patch.py \
  tests/model_executor/model_loader/test_checkpoint_weight_patch.py
```

The focused CPU test covers the dense seed, checkpoint-global sparse indices, TP-local slicing, packed Q/K destinations, repeated patch ordering, and restoration of `Tensor.copy_` after the supported path.

The companion VERL integration tests additionally cover:

- Qwen3-30B-A3B TP8+EP8 sparse-vs-dense and restore comparisons;
- real CUDA IPC/ZMQ transfer through the companion VERL adapter;
- a two-node full-NCCL versus `delta_sharded` weight-sync comparison.

### Focused repository test

- Pytest: **1 passed**.
- Ruff check: **passed**.
- Ruff format check: **passed**.

### TP8+EP8 integration correctness

Qwen3-30B-A3B, BF16, vLLM TP8+EP8:

- 8/8 workers passed.
- 32/32 selected runtime tensors matched the dense-load target with `torch.equal` after sparse apply.
- 32/32 tensors returned to the checkpoint baseline after restore.
- The probe covered packed QKV, norm, routed-expert `w13`, and routed-expert `w2` mappings.

This is a representative runtime-tensor mapping probe, not a full-model digest or a phase-for-phase training-trajectory comparison.

### Companion VERL integration performance

These are end-to-end results from the companion VERL integration, not a microbenchmark of this helper alone.

| mode | steady samples | mean | median | range |
|---|---:|---:|---:|---:|
| Full NCCL | 11 | 51.814 s | 51.812 s | 51.457-52.051 s |
| `delta_sharded` + vLLM | 11 | 7.200 s | 7.740 s | 5.943-8.686 s |

The ratio of means is **7.196x**; the conservative observed steady ratio is **5.924x**. The mean changed ratio was **0.646%**, and the mean sparse payload was **1129.4 MiB**. The one-time dense seed took **51.9 s** and is excluded from steady-state statistics.

Environment: two nodes with 8 H20 GPUs per node; Qwen3-30B-A3B; FSDP1 trainer on one node; vLLM TP8+EP8 Triton-MoE rollout on the other; BF16; VERL V1 `separate_async`; NCCL `NET/Socket` over `eth0`; 12 delta and 12 full steps, with steps 2-12 used for the primary statistics.

The two async jobs did not produce phase-aligned reward trajectories. The performance claim is limited to sustained weight-sync time; correctness comes from the independent bit-exact probe.

## AI assistance disclosure

OpenAI Codex was used for code exploration, implementation assistance, testing, and review.

### CODE REVIEW

**Review by aoshen02 — APPROVED** [review]

File: vllm/model_executor/model_loader/checkpoint_weight_patch.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+_DEFAULT_PATCH_CHUNK_BYTES = 512 << 20
+
+__all__ = [
+    "CheckpointWeightPatch",
+    "load_checkpoint_weight_patches",
+]
+
+
+@dataclass(frozen=True)
+class CheckpointWeightPatch:
```
Comment (Isotr0py) [inline vllm/model_executor/model_loader/checkpoint_weight_patch.py]:
```suggestion
class CheckpointWeightMetadata(NamedTuple):
```
I think metadata will be more proper here.

**Review by Isotr0py — APPROVED** [review]

File: vllm/model_executor/model_loader/checkpoint_weight_patch.py (hunk 1 above)
Comment (ShuoleiWang) [reply vllm/model_executor/model_loader/checkpoint_weight_patch.py]: Done
